xroad-catalog-collector/src/main/java/org/niis/xroad/catalog/collector/tasks/ListClientsTask.java

```diff
@@ -0,0 +1,127 @@
+import org.niis.xroad.catalog.collector.wsimport.ClientType;
+import org.niis.xroad.catalog.collector.wsimport.XRoadObjectType;
+import org.niis.xroad.catalog.persistence.entity.ErrorLog;
+import org.niis.xroad.catalog.persistence.entity.Member;
+import org.niis.xroad.catalog.persistence.entity.MemberId;
+import org.niis.xroad.catalog.persistence.entity.Subsystem;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+@Slf4j
+@Component
+public class ListClientsTask implements Runnable {
+
+    private final TaskPoolConfiguration taskPoolConfiguration;
+    private final CatalogService catalogService;
+    private final Queue<ClientType> listMethodsQueue;
+    private final NewMembersEventPublisher newMembersEventPublisher;
+
+    public ListClientsTask(CatalogService catalogService, TaskPoolConfiguration taskPoolConfiguration, Queue<ClientType> listMethodsQueue,
+                           NewMembersEventPublisher newMembersEventPublisher) {
+        this.taskPoolConfiguration = taskPoolConfiguration;
+        this.catalogService = catalogService;
+        this.listMethodsQueue = listMethodsQueue;
+        this.newMembersEventPublisher = newMembersEventPublisher;
+    }
+
+    public void run() {
+        log.info("Starting ListClientsTask");
+        if (CollectorUtils.isTimeBetweenHours(taskPoolConfiguration.getFlushLogTimeAfterHour(),
+                taskPoolConfiguration.getFlushLogTimeBeforeHour())) {
+            catalogService.deleteOldErrorLogEntries(taskPoolConfiguration.getErrorLogLengthInDays());
+        }
+
+        if (taskPoolConfiguration.isFetchRunUnlimited()
+                || CollectorUtils.isTimeBetweenHours(taskPoolConfiguration.getFetchTimeAfterHour(),
+                taskPoolConfiguration.getFetchTimeBeforeHour())) {
+            fetchClients();
+        }
+    }
+
+    private void fetchClients() {
+        String listClientsUrl = taskPoolConfiguration.getListClientsHost() + "/listClients";
+        try {
+            log.info("Getting client list from {}", listClientsUrl);
+            ClientList clientList = ClientListUtil.clientListFromResponse(listClientsUrl);
+            HashMap<MemberId, Member> m = populateMapWithMembers(clientList);
+            Set<Member> newMembers = catalogService.saveAllMembersAndSubsystems(m.values());
+
+            // We only fetch WSDL-s and REST services from subsystems
+            List<ClientType> subsystems = clientList.getMember().stream()
+                    .filter(client -> XRoadObjectType.SUBSYSTEM.equals(client.getId().getObjectType()))
+                    .toList();
+            listMethodsQueue.addAll(subsystems);
+
+            log.info("All subsystems ({}) sent to ListMethodsTask", subsystems.size());
+
+            newMembersEventPublisher.publishNewMembersEvent(newMembers.stream().map(Member::getMemberCode).toList());
```

Comment:
What happens here if there is no listener registered for the event (which I guess is the case if `fi` is not enabled)?

I have a feeling this might lead to some internal queue growing until we run out of memory.

Reply:
@raits, honestly, I am not sure.

Going through [the documentations](https://docs.spring.io/spring-framework/reference/core/beans/context-introduction.html#context-functionality-events), it says:
> but note that, by default, event listeners receive events synchronously. This means that the publishEvent() method blocks until all listeners have finished processing the event.

I also tried reading the Spring Framework's code base and I found this class [SimpleApplicationEventMulticaster](https://github.com/spring-projects/spring-framework/blob/main/spring-context/src/main/java/org/springframework/context/event/SimpleApplicationEventMulticaster.java#L142-L159).   

I think that the behavior is based on instant notification and not a queue based one. They also mentioned somewhere that if a more complex solution is required, then Spring Integration should be used instead for a more effective Event based system.

What do you think?

Reply:
Based on the code it does looks like a NOOP when there are no listeners registered. So it should probably be ok, but I think it might be good to profile this just in case an see if memory usage starts creeping up.

Reply:
@raits, I created this simple task to test the memory usage of the collector service and made it to run every 250 milliseconds:
```java

public class TestingTask implements Runnable {
    @Autowired
    private final NewMembersEventPublisher newMembersEventPublisher;

    static final String MEM_NAME = "TestingTask %s";
    static int count = 1;

    @Override
    public void run() {
        log.info("TestingTask is running");
        newMembersEventPublisher.publishNewMembersEvent(Set.of(MEM_NAME.formatted(count++)));
    }
}
```

My findings are that when I profile the application, memory heap size increases by 4 MB every 38 seconds.  
When I stop recording and just monitor the memory usage, heap size increases by 4 MB, but almost every 1 minute.

Using the following event listener:
```java

public class TestEventListener {

    private List<String> newMembersCodes = new ArrayList<>();

    @EventListener
    public void handleNewMembersEvent(NewMembersEvent event) {
        log.info("New members detected: {} and are being handled by {}", event.newMembersCodes(), this);
        newMembersCodes.addAll(event.newMembersCodes());
    }
}
```


I tried enabling this listener, disabling it, adding the member codes to a list (as per the code above) to keep it in memory and affect the heapsize, or just printing it in logs without storing it in the list.  
All had the same memory and heap size. Nothing has changed at all.
I also kept the application going for half an hour to make sure, but no difference.